velox/exec/OutputBuffer.h

```diff
@@ -124,10 +184,58 @@ class DestinationBuffer {
   // The sequence number of the first item to pass to 'notify'.
   int64_t notifySequence_{0};
   uint64_t notifyMaxBytes_{0};
+  BufferStats stats_;
 };
 
 class Task;
 
+/// The stats of an OutputBuffer.
+struct OutputBufferStats {
+  /// BufferState indicates the OutputBuffer's current state.
+  enum class BufferState {
+    /// The OutputBuffer is open and ready to receive data. Initial state.
+    kOpen,
+    /// The OutputBuffer would not accept more buffers.
+    kNoMoreBuffers,
+    /// The OutputBuffer has received all input data, and some data is still
+    /// to be sent.
+    kFlushing,
+    /// The OutputBuffer has already received and sent all the input data.
+    kFinished,
+    /// The OutputBuffer has failed.
+    kFailed
+  };
+
+  /// The type name of the OutputBuffer.
+  std::string type;
+  /// The state of the OutputBuffer.
+  BufferState state{BufferState::kOpen};
+  /// If the OutputBuffer can add more buffers.
+  bool canAddBuffers{true};
+  /// If the outputBuffer can receive more input pages.
+  bool canAddPages{true};
+  /// The bufferStats of each destination buffer.
+  std::vector<BufferStats> buffers;
+  /// Total byte num of data buffered in all destination buffers.
+  int64_t totalBytesBuffered{0};
```

Comment:
It is the case that totalAdded = totalBuffered + totalSent? If this is so, then perhaps, we can remove totalAdded as these are redundant.

Reply:
Yes, totalAdded = totalBuffered + totalSent, but I tend to keep the "redundant" totalAdded, because the 'added' -> 'buffered' -> 'sent' are three different phases for an enqueued page, and keep each phase recorded separately is **more straightforward and easier to understand**.

Well, if you insist to remove the redundancy, I could remove the field then. Also note that actually all the `totalXXX` fields in `OutputBufferStats` are simply accumulations of corresponding `BufferStats`, which could be considered as redundant as well. So if we are going to reduce all the redundancy, all the `totalXXX` fields could be removed.

Reply:
@HolyLow These are good points. It would be nice to reduce redundancy. Thanks.

Reply:
done.